pkgs/development/compilers/llvm/15/llvm/default.nix

```
@@ -0,0 +1,299 @@
+  || stdenv.isAarch64 # broken for Ampere eMAG 8180 (c2.large.arm on Packet) #56245
+  || stdenv.isAarch32 # broken for the armv7l builder
+)
+, enablePolly ? false
+} @args:
+
+let
+  inherit (lib) optional optionals optionalString;
+
+  # Used when creating a version-suffixed symlink of libLLVM.dylib
+  shortVersion = with lib;
+    concatStringsSep "." (take 1 (splitString "." release_version));
+
+in stdenv.mkDerivation (rec {
+  pname = "llvm";
+  inherit version;
+
+  src = runCommand "${pname}-src-${version}" {} (''
+    mkdir -p "$out"
+    cp -r ${monorepoSrc}/cmake "$out"
+    cp -r ${monorepoSrc}/${pname} "$out"
+    cp -r ${monorepoSrc}/third-party "$out"
+  '' + lib.optionalString enablePolly ''
+    cp -r ${monorepoSrc}/polly "$out/llvm/tools"
+  '');
+
+  sourceRoot = "${src.name}/${pname}";
+
+  outputs = [ "out" "lib" "dev" "python" ];
+
+  nativeBuildInputs = [ cmake ninja python3 ]
+    ++ optionals enableManpages [ python3.pkgs.sphinx python3.pkgs.recommonmark ];
+
+  buildInputs = [ libxml2 libffi ]
+    ++ optional enablePFM libpfm; # exegesis
+
+  propagatedBuildInputs = [ ncurses zlib ];
+
+  checkInputs = [
+    which
+    python3.pkgs.psutil # !!! lit still doesn't seem to find this; leads to some
+                        # tests being skipped
+  ];
+
+  patches = [
+    ./gnu-install-dirs.patch
+  ] ++ lib.optional enablePolly ./gnu-install-dirs-polly.patch;
+
+  postPatch = optionalString stdenv.isDarwin ''
+    substituteInPlace cmake/modules/AddLLVM.cmake \
+      --replace 'set(_install_name_dir INSTALL_NAME_DIR "@rpath")' "set(_install_name_dir)" \
+      --replace 'set(_install_rpath "@loader_path/../''${CMAKE_INSTALL_LIBDIR}''${LLVM_LIBDIR_SUFFIX}" ''${extra_libdir})' ""
+  '' + ''
+    # FileSystem permissions tests fail with various special bits
+    substituteInPlace unittests/Support/CMakeLists.txt \
+      --replace "Path.cpp" ""
+    rm unittests/Support/Path.cpp
+    substituteInPlace unittests/IR/CMakeLists.txt \
+      --replace "PassBuilderCallbacksTest.cpp" ""
+    rm unittests/IR/PassBuilderCallbacksTest.cpp
+    rm test/tools/llvm-objcopy/ELF/mirror-permissions-unix.test
+  '' + optionalString stdenv.hostPlatform.isMusl ''
+    patch -p1 -i ${../../TLI-musl.patch}
+    substituteInPlace unittests/Support/CMakeLists.txt \
+      --replace "add_subdirectory(DynamicLibrary)" ""
+    rm unittests/Support/DynamicLibrary/DynamicLibraryTest.cpp
+    # valgrind unhappy with musl or glibc, but fails w/musl only
+    rm test/CodeGen/AArch64/wineh4.mir
+  '' + optionalString stdenv.hostPlatform.isAarch32 ''
+    # skip failing X86 test cases on 32-bit ARM
+    rm test/DebugInfo/X86/convert-debugloc.ll
+    rm test/DebugInfo/X86/convert-inlined.ll
+    rm test/DebugInfo/X86/convert-linked.ll
+    rm test/tools/dsymutil/X86/op-convert.test
+    rm test/tools/gold/X86/split-dwarf.ll
+    rm test/tools/llvm-dwarfdump/X86/prettyprint_types.s
+    rm test/tools/llvm-dwarfdump/X86/simplified-template-names.s
```

```suggestion
    rm test/tools/llvm-dwarfdump/X86/simplified-template-names.s
    rm test/CodeGen/RISCV/rv32zbp.ll
    rm test/CodeGen/RISCV/rv64zbp.ll
```

These fail on armv7l-linux:
```
llvm> ********************
llvm> ********************
llvm> Failed Tests (2):
llvm>   LLVM :: CodeGen/RISCV/rv32zbp.ll
llvm>   LLVM :: CodeGen/RISCV/rv64zbp.ll
llvm> Testing Time: 155.73s
llvm>   Skipped          :    20
llvm>   Unsupported      :  1596
llvm>   Passed           : 47426
llvm>   Expectedly Failed:   161
llvm>   Failed           :     2
llvm> 1 warning(s) in tests
llvm> FAILED: CMakeFiles/check-all /build/llvm-src-15.0.2/llvm/build/CMakeFiles/check-all
llvm> cd /build/llvm-src-15.0.2/llvm/build && /nix/store/pqz0gq3pslichbh57qy4z3cwj4rm68fb-python3-3.10.7/bin/python3.10 /build/llvm-src-15.0.2/llvm/build/./bin/llvm-lit -svj80 --no-progress-bar /build/llvm-src-15.0.2/llvm/build/utils/lit /build/llvm-src-15.0.2/llvm/build/test
llvm> ninja: build stopped: subcommand failed.
```

---

Thanks for testing this! Unfortunately I don't have access to a native armv7 machine so I can't dig into this myself — do you have any information or insight into what is causing these tests to fail?

I'd like to, at the very least, leave some context about why we're disabling these tests in the comments.

---

Here is the full build log: [llvm_15.log](https://github.com/NixOS/nixpkgs/files/9823656/llvm_15.log)
You can also use community aarch64 NixOS box to build for armv7l-linux (see https://github.com/nix-community/aarch64-build-box/pull/133)

---

Probably related: https://github.com/llvm/llvm-project/issues/50553

---

If I'm understanding correctly filecheck is complaining about seeing:
```risc-v
  rev8 a0, a0
  rori a0, a0, 32
  ret
```

instead of:
```risc-v
  rev8.w a0, a0
  ret
```
when ZBP is enabled?
(and something similarly mismatched for `bswap_rotl_i64`)

It's kind of curious that we don't seem to error on this on aarch64/x86_64 but, given that it looks like those tests (along with support for that extension in general) [have been dropped in LLVM 16](https://github.com/llvm/llvm-project/commit/182aa0cbe0cd86dbf91e9d0e647eec736461f0b7#diff-111093152696dbf9fa35b4948b543cdd57b42b149aead45d7864f80f9c0cafa3) it's probably not worth looking into further.

---

@misuzu thanks for looking into this; I'll commit your suggestion and add a note saying that we can drop this change in future `llvmPackages`